---
title: "phylociraptor report"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
---
## {.tabset .tabset-fade}

```{r setup, echo=FALSE, message=FALSE}
library(kableExtra)
#library(formattable)
library(dplyr)
library(knitr)
#library(tidyverse)
library(ggplot2)
library(viridis)
library(readr)
library(plotly)
library(stringr)
library(data.table)
#library(ggpubr)

parsimony_cutoff <- 3


# function to convert seconds to better formated time:
# from here: https://stackoverflow.com/a/51156986

dhms <- function(t){
    paste(t %/% (60*60*24), "days"
         ,paste(formatC(t %/% (60*60) %% 24, width = 2, format = "d", flag = "0")
               ,formatC(t %/% 60 %% 60, width = 2, format = "d", flag = "0")
               ,formatC(t %% 60, width = 2, format = "d", flag = "0")
               ,sep = ":"
               )
         ,sep=" ")
}

```

```{r checkinput_general, echo=FALSE}

# set input file names:

busco_overview_image <- ""

# orthology statistics file:
busco_overview_file <- "../results/orthology/busco/busco_set/dataset.cfg"
busco_summary_file <- "../results/statistics/busco_summary.txt"
orthology_filtering_genomes_file <- "../results/statistics/orthology_filtering_genomes_statistics.txt"
orthology_filtering_genes_file <- "../results/statistics/orthology_filtering_gene_statistics.txt"


#  alignment statistics information
alignment_statistics_file <- "../results/statistics/statistics_alignments.txt"
trimmed_alignment_statistics_file <- "../results/statistics/statistics_trimmed.txt"
filtered_alignment_statistics_file <- "../results/statistics/statistics_filtered.txt"
filtered_gene_information_file <- "../results/statistics/alignment_filter_information.txt"

align_trim_overview_statistics_file <- "../results/statistics/align_trim_overview_statistics.txt"

downloaded_genomes_statistics_file <- "../results/statistics/downloaded_genomes_statistics.txt"
failed_genome_downloads_file <- "../results/downloaded_genomes/not_downloaded.txt"
successfull_genome_downloads_file <- "../results/downloaded_genomes/successfully_downloaded.txt"
local_species_file <- "../results/statistics/local_species.txt"

best_models_file <- "../results/modeltest/best_models.txt"

tree_statistics_file <- "../results/statistics/tree_statistics.txt"
speciestree_statistics_file <- "../results/statistics/speciestree_statistics.txt"
runlog_file <- "../results/statistics/runlog.txt"

#read benchmarking results

benchmarkfile_run_busco <- "../results/statistics/benchmark_all_busco_runs.bench"


# global variables:

base_size <- 12
```


### Genomes

```{r genomes, echo=FALSE}
if (file.exists(failed_genome_downloads_file))
{
  if (file.size(failed_genome_downloads_file) > 0) {
    failed_sp <- read.table(failed_genome_downloads_file, header=F)
    failed <- paste(failed_sp$V1, sep="\n")
  } else { failed <- numeric()}
} else {
  failed <- numeric()
}

if (file.exists(successfull_genome_downloads_file)){
  #print(file.info(successfull_genome_downloads_file)$size)
  if (file.size(successfull_genome_downloads_file) > 0) {
    success_sp <- read.table(successfull_genome_downloads_file, header=F)
    success <- paste(success_sp$V1, sep="\n")
  } else {success <- numeric()}
} else {
  success <- numeric()
}

if (file.exists(local_species_file)){
  if (file.size(local_species_file) > 0) {
  local_sp <- read.table(local_species_file, header=F)
  local <- paste(local_sp$V1, sep="\n")
  } else {local <- numeric()}
} else {
  local <- numeric()
}

total <- length(failed) + length(success) + length(local)

```


**Location of genomes files:** 

```
results/assemblies
```

**Total number of included genomes:** `r total`

**Locally provided genomes:** `r length(local)`

**Successfully downloaded genomes:** `r length(success)`

**Failed species downloads:** `r length(failed)`

**Not downloaded:**<br />`r failed`

**Information on successfully downloaded genomes:**
```{r genome_stats, echo=FALSE, out.height=500}
info <- ""
if (file.exists(downloaded_genomes_statistics_file))
{
  if (file.size(downloaded_genomes_statistics_file) != 0) {
      data <- read.csv(downloaded_genomes_statistics_file, sep="\t", header=T)
  } else {
    info <- "*No information about downloaded genomes found. Maybe you only have local species?*"
  }
  
  data %>% mutate(ftp_path = cell_spec("download", "html", link = ftp_path)) %>% kbl(escape=F) %>% kable_paper("striped", full_width = F) %>% scroll_box(width = "100%", height = "600px")
} else {
  info <- "*No information about downloaded genomes found. Did you run phylociraptor setup ?*"
}

```
`r info`


### Orthology

```{r, echo=FALSE}

if (file.exists(busco_overview_file)){
  busco_overview <- read.table(busco_overview_file, sep="=", header=F)
  busco_set <- busco_overview$V2[busco_overview$V1 == "name"]
  nbuscos <- busco_overview$V2[busco_overview$V1 == "number_of_BUSCOs"]
  
  
} else {
  busco_set <- "unknown (maybe BUSCO set was not downloaded yet)"
  nbuscos <- "unknown (maybe BUSCO set was not downloaded yet)"
}

if (file.exists(benchmarkfile_run_busco)) {
  busco_bench_df <- read.table(benchmarkfile_run_busco, sep="\t", header=F)
  colnames(busco_bench_df) <- c("job", "seconds", "time", "max_vss", "max_vms", "max_uss", "max_pss", "io_in", "io_out", "mean_load")
  busco_bench_df$minutes <- busco_bench_df$seconds/60
  busco_bench_df$job <- str_replace(busco_bench_df$job, "run_busco_", "")
  busco_bench_df$job <- str_replace(busco_bench_df$job, ".txt", "")
} else {
  busco_bench_df <- "not_found"
}

if (file.exists(orthology_filtering_genomes_file)) {
  ortho_filter_genomes <- read.table(orthology_filtering_genomes_file, sep="\t", header=F)
  colnames(ortho_filter_genomes) <- c("sample", "status", "completeness", "cutoff")
  failed_genomes <- ortho_filter_genomes$sample[ortho_filter_genomes$status == "FAILED"]
  cutoff <- ortho_filter_genomes$cutoff[1]
} else {
  failed_genomes <- numeric()
  cutoff <- numeric()
}
if (file.exists(orthology_filtering_genes_file)) {
  ortho_filter_genes <- read.table(orthology_filtering_genes_file, sep="\t", header=F)
  colnames(ortho_filter_genes) <- c("gene", "status", "nseqs", "cutoff")
  failed_genes <- ortho_filter_genes$gene[ortho_filter_genes$status == "FAILED"]
  cutoff_gene <- ortho_filter_genes$cutoff[1]
} else {
  failed_genes <- numeric()
  cutoff_gene <- numeric()
}

```
Orthologous genes were inferred using BUSCO.

**Location of BUSCO results:** 

```
results/orthology/busco
```

**Used BUSCO set:** `r busco_set`

**Total number of BUSCO genes in BUSCO set:** `r nbuscos`  
**Number of BUSCO genes recovered in too few (`r cutoff_gene`) samples:** `r length(failed_genes)`  
**BUSCO genes below threshold:**  
`r failed_genes`

**Number of samples with single-copy BUSCO score below threshold (`r cutoff`):** `r length(failed_genomes)`  
**Genomes below threshold:**  
`r failed_genomes`

**Runtime Statistics:**  

|        Total (cumulative) runtime:
|        `r if (busco_bench_df != "not_found") { dhms(sum(busco_bench_df$seconds)) }`

```{r, echo=F, message=F, warning=F, fig.width=8}
if (busco_bench_df != "not_found") {
  nrows <- nrow(busco_bench_df)
  busco_bench_df <- busco_bench_df[ order(row.names(busco_bench_df)), ]
  p <- ggplot(busco_bench_df, aes(x=minutes, y=job)) +geom_bar(stat="identity") +theme(axis.ticks.x = element_blank(),axis.ticks.y = element_blank(), panel.grid=element_blank(), panel.background = element_blank(), axis.text.x = element_text(size = base_size *0.8, hjust = 0,vjust=1, colour = "grey50"))
  
  #+theme_bw()+ theme(axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank())
  
  }

```


```{r busco_stats, echo=FALSE, message=F, fig.width=8, warning=F}
#if (file.exists("../results/statistics/busco_summary.txt"))
if (file.exists(busco_summary_file))
{
  data <- read.table(busco_summary_file, sep="\t", header=T)
  data <- data[ order(row.names(data)), ]
  data$total <- NULL
  data$complete <- NULL
  nrows <- nrow(data)
  mdata <- reshape2::melt(data, id.vars="species")

  colnames(mdata) <- c("species", "BUSCO category", "no. of genes")
  p2 <- ggplot(mdata, aes_string(fill="`BUSCO category`", x="`no. of genes`", y="`species`")) +geom_bar(position="stack", stat="identity")+theme(axis.ticks.x = element_blank(),axis.ticks.y = element_blank(), panel.grid=element_blank(), panel.background = element_blank(), axis.text.x = element_text(size = base_size *0.8, hjust = 0,vjust=1, colour = "grey50"),,axis.text.y=element_blank())+scale_fill_manual(values=c("#6198B8", "#E37332", "#C6E88B", "#BCB2CF"))#+ggtitle(paste("Overview of BUSCO results (total genes=", 

if (busco_bench_df != "not_found") {
  subplot(ggplotly(p, height=15*nrows), ggplotly(p2, height=15*nrows))
} else {
  p2 <- p2 + theme(axis.text.y=element_text(size = base_size *0.8, hjust = 0,vjust=1, colour = "grey50"))
  ggplotly(p2, height=15*nrows)
}
  
  #data %>% kbl(escape=F) %>% kable_paper("striped", full_width = F) %>% scroll_box(width = "100%", height = "600px") 
} else {
  cat("no information on BUSCO runs found. Did you run phylociraptor orthology?")
}


```

```{r, include=TRUE, fig.align="center", echo=F}
#out.width="1.0\\linewidth"
if (file.exists(busco_overview_image)) {
  include_graphics(busco_overview_image)
} 
#else {
#  print("Plot file not found.")
#}

```

### Alignments

**Location of alignment files:** 
```
results/alignments/full
```


```{r stats_alignments, echo=FALSE}
dirs <- list.dirs("../results/statistics/", recursive=F)
dirs <- grep("align-*", dirs, value = TRUE)

read_my_overview <-function(dat) {
	return(read.csv(dat,header=F,sep="\t"))
}

if (length(dirs) != 0) {
	dat <- list()
	i <- 1
	for (dir in dirs) { # this code needs to be expanded and the dataframe integrated into the one created in the next chunk
		files <- list.files(path=dir, pattern="*overview*", full.names=T)
		data <- do.call(rbind,lapply(files,read_my_overview))	
		dat[[i]] <- data
		i <- i + 1
	}
	
	data_combined_overview <- do.call(cbind, dat)
	#print(data_combined_overview)
}
```

```{r stats_alignments2, echo=FALSE, results="asis"} 
if (length(dirs)==0) {

cat("<br><b> Alignment statistic files not found. Did you run phylociraptor align?</b>\n")

} else {
	
	read_my_csv <-function(dat) {
		return(read.csv(dat,header=T,sep="\t"))
	}
	
	dat <- list()
	i <- 1
	for (dir in dirs) {
		files <- list.files(path=dir, pattern="*statistics*", full.names=T)
		data <- do.call(rbind,lapply(files,read_my_csv))	
		data[,6] <- cell_spec(data[,6], background = ifelse(data[,6] < 210, "red", ""))
		if (i ==1) {
			data <- data[,c(1,4:8)]
		} else{
			data <- data[,c(4:8)]
		}
		rows <- which(data$nparsimony < 10)
		data <- as.data.frame(data)
		dat[[i]] <- data 
		i <- i + 1
	}
	
	data_combined <- do.call(cbind, dat)
	rownames(data_combined) <- data_combined$alignment
	data_combined$alignment <- NULL
	
	get_headers <- function(dirs) {
	
	values <- c(1)
	names <- c(" ")
	
	for (dir in dirs) {
		n <- tail(str_split(dir, "-")[[1]], n=1)
		values <- c(values, 5)
		names <- c(names, n)
	
	}
	names(values) <- names
	return(values)
	}
	
	
	data_combined %>% kbl(escape=F) %>% add_header_above(get_headers(dirs)) %>% kable_paper("striped", full_width = F) %>% scroll_box(width = "100%", height = "600px") 
	
}

#if (length(dirs) > 0) {
#cat(paste0("<br><b>Alignment statistics for ", dirs[1] ,"</b>"))
#dat[[1]]
#}
#
#if (length(dirs) > 1) {
#cat(paste0("<br><b>Alignment statistics for ", dirs[2] ,"</b>"))
#dat[[2]]
#}
#if (file.exists(alignment_statistics_file) == TRUE) {
#    rows <- which(data$nparsimony < 10)
#    data %>% kbl(escape=F) %>% kable_paper("striped", full_width = F) %>% row_spec(rows, bold = T, color="white", background="red") %>% scroll_box(width = "100%", height = "600px")
#} else {
#  print("Alignment statistics file not found. Did you run phylociraptor align?")
#}
#%>% row_spec(1:nrow(data), color = ifelse(data$nparsimony < parsimony_cutoff, "red", "black")) 
  #column_spec(1, color = "white",background = spec_color(data$nparsimony < 10 ,begin=0, end = 0.6, option="B", direction=1))
```

### Trimming

**Location of trimmed alignment files:** 
```
results/alignments/trimmed
```

```{r stats_trimmed, echo=FALSE}
dirs <- list.dirs("../results/statistics/", recursive=F)
dirs <- grep("/trim-*", dirs, value = TRUE)


if (file.exists(align_trim_overview_statistics_file) == TRUE) {
  ov <- read.table(align_trim_overview_statistics_file, sep="\t", header=F)
  tr_method <- ov[3,2]
  tr_params <- ov[4,2]
  parsimony_cutoff <- as.numeric(ov[5,2])
} else {
  tr_method <- "unknown (check if phylocraptor -m align has finished successfully)."
  tr_params <- numeric()
  parsimony_cutoff <- 0
}

if (file.exists(trimmed_alignment_statistics_file) == TRUE) {
  data <- read.table(trimmed_alignment_statistics_file, sep="\t", header=T)
  few_npars <- data$alignment[data$nparsimony < parsimony_cutoff]
} else {
  few_npars <- numeric()
}
```

**Trimming method:** `r tr_method` `r tr_params`  
**Total number of alignments after trimming:** `r nrow(data)`  
**Alignments with less than `r parsimony_cutoff` parsimony informative sites:** `r length(few_npars)`  

```{r stats_trimmed2, echo=FALSE, warning=FALSE}


if (length(dirs)==0) {
cat("<br><b> Trimming statistic files not found. Did you run phylociraptor filter-align?</b>\n")

} else {
	read_my_csv <-function(dat) {
		return(read.csv(dat,header=T,sep="\t"))
	}
	
	dat <- list()
	i <- 1
	for (dir in dirs) {
		files <- list.files(path=dir, pattern="*statistics*", full.names=T)
		data <- do.call(rbind,lapply(files,read_my_csv))	
		data[,6] <- cell_spec(data[,6], background = ifelse(data[,6] < 210, "red", ""))
		data <- data[,c(1,4:8)]
		data <- as.data.frame(data)
		dat[[i]] <- data 
		i <- i + 1
	}
	
	multi_merge <- function(d) {
		df <- d[[1]]
		if (length(d) >= 2) {
			for (i in 2:length(dat)) {
				df <- merge(df, d[[i]], by="alignment", all=T, suffixes=c("",""))
				#df <- full_join(df, d[[i]], by="alignment", suffix = c(".x", ".y"))
			}
		}
		return(df)
	}

	#data_combined <- merge(dat[[1]], dat[[2]], all=T, by="alignment")
	data_combined <- do.call(multi_merge, list(dat))
	data_combined[is.na(data_combined)] = ""


	#rownames(data_combined) <- data_combined$alignment
	#data_combined$alignment <- NULL
	#print(data_combined %>% kbl())	
	
	get_headers_trimmer <- function(dirs) {
	
	values <- c(1)
	names <- c(" ")
	
	for (dir in dirs) {
		vec <- str_split(dir, "-")[[1]]
		n <- vec[length(vec)-1]
		
		values <- c(values, 5)
		names <- c(names, n)
	
	}
	names(values) <- names
	return(values)
	}
		
	get_headers_aligner <- function(dirs) {
	
	values <- c(1)
	names <- c(" ")
	
	for (dir in dirs) {
		n <- tail(str_split(dir, "-")[[1]], n=1)
		values <- c(values, 5)
		names <- c(names, n)
	
	}
	names(values) <- names
	return(values)
	}
	
	data_combined %>% kbl(escape=F) %>% add_header_above(get_headers_aligner(dirs)) %>% add_header_above(get_headers_trimmer(dirs)) %>% kable_paper("striped", full_width = F) %>% scroll_box(width = "100%", height = "600px") 
	
}

#if (file.exists(trimmed_alignment_statistics_file) == TRUE) {
#rows <- which(data$nparsimony < parsimony_cutoff)
##data$nparsimony <- cell_spec(data$nparsimony, color = ifelse(data$nparsimony < parsimony_cutoff, "red", "black"))
##data$alignment <- cell_spec(data$alignment, color = ifelse(data$nparsimony[data$alignment] < parsimony_cutoff, "red", "black"))
##print(rows)
#data %>% kbl(escape=F) %>% kable_paper("striped", full_width = F) %>% row_spec(rows, bold = T, color="white", background="red") %>% scroll_box(width = "100%", height = "600px")
#} else {
#  print("Trimming statistics file not found. Did you run phylociraptor align?")
#}
##%>% row_spec(1:nrow(data), color = ifelse(data$nparsimony < parsimony_cutoff, "red", "black"))
  
  
  
  #column_spec(1, color = "white",background = spec_color(data$nparsimony < 10 ,begin=0, end = 0.6, option="B", direction=1))
```


### Filtering

```{r stats_filtered, echo=FALSE}
dirs <- list.dirs("../results/statistics/", recursive=F)
dirs <- grep("/filter-*", dirs, value = TRUE)

if (file.exists(filtered_alignment_statistics_file) == TRUE) {
  data <- read.table(filtered_alignment_statistics_file, sep="\t", header=T)
  few_npars <- data$alignment[data$nparsimony < parsimony_cutoff]
} else {
  few_npars <- numeric()
}
if (file.exists(filtered_gene_information_file) == TRUE) {
  data_filtered <- read.table(filtered_gene_information_file, sep="\t", header=F)
  colnames(data_filtered) <- c("alignment", "status")
  uninformative_alignments <- data_filtered$alignment[data_filtered$status == "NOT_INFORMATIVE"]
  too_few_seqs_alignments <- data_filtered$alignment[data_filtered$status == "TOO_FEW_SEQUENCES"]
  dup_seqs_removed <- data_filtered$alignment[data_filtered$status == "REMOVE_DUPLICATED_SEQUENCES"]
  dup_align_removed <- data_filtered$alignment[data_filtered$status == "REMOVED"]
} else {
  uninformative_alignments <- numeric()
  too_few_seqs_alignments <- numeric()
  dup_seqs_removed <- numeric()
  dup_align_removed <- numeric()
}

```

**Location of filtered alignment files:** 
```
results/alignments/filtered
```
<!---
**Total number of alignments after filtering:** `r nrow(data)`  
**Number of discarded alignments:** `r length(uninformative_alignments)+length(too_few_seqs_alignments)+length(dup_align_removed)`  

**The following alignments were discarded:**  

_Too few parsimony informative sites:_  
`r if (length(uninformative_alignments) == 0){"none"}else{uninformative_alignments}`

_Less then three sequences after trimming:_  
`r if (length(too_few_seqs_alignments) == 0){"none"}else{too_few_seqs_alignments}`

_Removed alignments due to duplicated sequences:_  
`r if (length(dup_align_removed) == 0){"none"}else{dup_align_removed}`

_Alignments with removed (duplicated) sequences:_  
`r if (length(dup_seqs_removed)==0){"none"}else{dup_seqs_removed}`

--->

```{r stats_filtered2, echo=FALSE, warning=FALSE}

if (length(dirs)==0) {
cat("<br><b> Filtering statistic files not found. Did you run phylociraptor filter-align?</b>\n")

} else {
	read_my_csv <-function(dat) {
		return(read.csv(dat,header=T,sep="\t"))
	}
	
	dat <- list()
	i <- 1
	for (dir in dirs) {
		files <- list.files(path=dir, pattern="*statistics*", full.names=T)
		data <- do.call(rbind,lapply(files,read_my_csv))	
		data[,6] <- cell_spec(data[,6], background = ifelse(data[,6] < 210, "red", ""))
		data <- data[,c(1,4:8)]
		data <- as.data.frame(data)
		dat[[i]] <- data 
		i <- i + 1
	}
	
	multi_merge <- function(d) {
		df <- d[[1]]
		if (length(d) >= 2) {
			for (i in 2:length(dat)) {
				df <- merge(df, d[[i]], by="alignment", all=T, suffixes=c("",""))
				#df <- full_join(df, d[[i]], by="alignment", suffix = c(".x", ".y"))
			}
		}
		return(df)
	}

	#data_combined <- merge(dat[[1]], dat[[2]], all=T, by="alignment")
	data_combined <- do.call(multi_merge, list(dat))
	data_combined[is.na(data_combined)] = ""

	#rownames(data_combined) <- data_combined$alignment
	#data_combined$alignment <- NULL
	#print(data_combined %>% kbl())	
	
	get_headers_trimmer <- function(dirs) {
	
	values <- c(1)
	names <- c(" ")
	
	for (dir in dirs) {
		vec <- str_split(dir, "-")[[1]]
		n <- vec[length(vec)-1]
		
		values <- c(values, 5)
		names <- c(names, n)
	
	}
	names(values) <- names
	return(values)
	}
		
	get_headers_aligner <- function(dirs) {
	
	values <- c(1)
	names <- c(" ")
	
	for (dir in dirs) {
		n <- tail(str_split(dir, "-")[[1]], n=1)
		values <- c(values, 5)
		names <- c(names, n)
	
	}
	names(values) <- names
	return(values)
	}
	
	data_combined %>% kbl(escape=F) %>% add_header_above(get_headers_aligner(dirs)) %>% add_header_above(get_headers_trimmer(dirs)) %>% kable_paper("striped", full_width = F) %>% scroll_box(width = "100%", height = "600px") 
	
}
```

### Modeltesting

**Location of modeltesting files:** 
```
results/modeltest
```

```{r stats_modeltest, echo=FALSE, results="asis", out.width="50%", warning=F}
files <- list.files(path="../results/modeltest/", pattern ="best_models_*", recursive=F)
multi_merge <- function(d) {
	df <- d[[1]]
	if (length(d) >= 2) {
		for (i in 2:length(dat)) {
			df <- merge(df, d[[i]], by="alignment", all=T, suffixes=c("",""))
			#df <- full_join(df, d[[i]], by="alignment", suffix = c("", ""))
		}
	}
	return(df)
}
if (length(files) > 0) {
	i <- 1
	dat <- list()
	all_genes <- c()
	names <- c("alignment")
	for (file in files) {
		# get reusable and meaningful name
		title <- str_split(file, "_")[[1]][3:4]
		# use other stringsplit function which supports fixed=T
		title[2] <- strsplit(title[2], ".", fixed=T)[[1]][1]
		title <- paste(title, collapse="-")
 		names <- c(names, title)
		dat[[i]] <- read.table(paste0("../results/modeltest/", file), sep="\t", header=F)
  		colnames(dat[[i]]) <- c("alignment", "model")
  		all_genes <- c(all_genes,dat[[i]]$alignment)
		models <- table(dat[[i]]$model)
  		p <- ggplot(dat[[i]], aes(x=factor(1), fill=factor(model))) + geom_bar(stat="count") + coord_polar(theta="y") + theme_void() + scale_fill_viridis(title,discrete=T, option="B") 
  		print(p)
		i <- i + 1
	}
	data_combined <- do.call(multi_merge, list(dat))
	colnames(data_combined) <- names
	data_combined %>% kbl(escape=F) %>% kable_paper("striped", full_width = F)%>% scroll_box(width = "100%", height = "600px")	
}
```



### Tree

```{r stats_tree, echo=FALSE}
if (file.exists(tree_statistics_file) == TRUE) {
  tree_stats <- TRUE
  tree_log <- read_file(tree_statistics_file)
  cat(tree_log)
} else {
  tree_stats <- FALSE
}
```

`r if(!tree_stats){"**Tree statistics not found. Did you run phylociraptor tree?**"}`

### Speciestree

```{r stats_speciestree, echo=FALSE}
if (file.exists(speciestree_statistics_file) == TRUE) {
  tree_stats <- TRUE
} else {
  tree_stats <- FALSE
}
```

`r if(!tree_stats){"**Speciestree statistics not found. Did you run phylociraptor speciestree?**"}`

### Log

```{r, echo=FALSE}
if (file.exists(runlog_file))
{
  runlog <- read_file(runlog_file)
} else {
  runlog <- "Information not found. Did you run phylociraptor?\n"
}
cat(runlog)
```

